File: PathFindingFun Shared/World.swift
//
//  World.swift
//  PathFindingFun iOS
//
//  Created by Carl Wieland on 8/26/19.
//  Copyright © 2019 Datum Apps. All rights reserved.
//

import Foundation

typealias MapPoint = Point<Int>

class World {
    static let TileSize = 50
    public private(set) var nodes: [MapNode]

    let width: Int
    let height: Int

    init(width: Int, height: Int) {
        self.width = width
        self.height = height
        nodes = [MapNode](repeating: MapNode(), count: width * height)

    }

    func position(of point: MapPoint) -> CGPoint {

        var result = CGPoint(x: point.x * World.TileSize, y: 0)

        if (point.y & 1) != 0 {
            result.x += CGFloat(World.TileSize / 2)
        }
        result.y = CGFloat(point.y * World.TileSize)
        return result
    }

    func index(of pt: MapPoint) -> Int {
        return pt.y * width + pt.x
    }

    func node(at pt: MapPoint) -> MapNode {
        return nodes[index(of: pt)]
    }

    func object(at pt: MapPoint) -> MapObject? {
        return node(at: pt).mapObject
    }

    func setObject(_ obj: MapObject, at pt: MapPoint) {
        nodes[index(of: pt)].mapObject = obj
    }

    func setPointRoad(at pt: MapPoint, direction: Direction, type: Road.RoadType) {
        if direction.isEastOrSouth {
            setNodeRoad(at: pt, in: direction.reversed, to: type)
        } else {
            setNodeRoad(at: neighbor(of: pt, direction: direction), in: direction, to: type)
        }

    }

    func setNodeRoad(at pt: MapPoint, in direction: Direction, to type: Road.RoadType) {
        assert(direction.isWestOrNorth)
        nodes[index(of: pt)].roads[Int(direction.rawValue)] = type
    }

    final func distance( x1: Int, y1: Int, x2: Int, y2: Int) -> Int {
        var dx = ((x1 - x2) * 2) + (y1 & 1) - (y2 & 1)
        var dy = ((y1 > y2) ? (y1 - y2) : (y2 - y1)) * 2

        if dx < 0 {
            dx = -dx
        }

        if dy > Int(height) {
            dy = (Int(height) * 2) - dy
        }

        if dx > Int(width) {
            dx = (Int(width) * 2) - dx
        }

        dx -= dy / 2

        return ((dy + (dx > 0 ? dx : 0)) / 2)
    }

    final func distance(from p1: MapPoint, to p2: MapPoint) -> Int {
        return distance(x1: p1.x, y1: p1.y, x2: p2.x, y2: p2.y)
    }

    final func neighbor(of pt: MapPoint, direction: Direction) -> MapPoint {
        /*  Note that every 2nd row is shifted by half a triangle to the left, therefore:
         Modifications for the dirs:
         current row:    Even    Odd
                      W  -1|0   -1|0
         D           NW  -1|-1   0|-1
         I           NE   0|-1   1|-1
         R            E   1|0    1|0
                     SE   0|1    1|1
                     SW  -1|1    0|1
         */

        var res = MapPoint(x:0, y: 0)
        switch direction {
        case .west: // -1|0   -1|0
            res.x = ((pt.x == 0) ? width : pt.x) - 1
            res.y = pt.y
        case .northwest: // -1|-1   0|-1
            res.x = (pt.y & 1) != 0 ? pt.x : (((pt.x == 0) ? width : pt.x) - 1)
            res.y = ((pt.y == 0) ? height : pt.y) - 1
        case .northeast: // 0|-1  -1|-1
            res.x = (!(pt.y & 1 != 0)) ? pt.x : ((pt.x == width - 1) ? 0 : pt.x + 1)
            res.y = ((pt.y == 0) ? height : pt.y) - 1
        case .east: // 1|0    1|0
            res.x = pt.x + 1
            if(res.x == width) {
                res.x = 0
            }
            res.y = pt.y
        case .southeast: // 1|1    0|1
            res.x = (!(pt.y & 1 != 0)) ? pt.x : ((pt.x == width - 1) ? 0 : pt.x + 1)
            res.y = pt.y + 1
            if(res.y == height) {
                res.y = 0
            }
        case .southwest: // 0|1   -1|1
            res.x = (pt.y & 1 != 0) ? pt.x : (((pt.x == 0) ? width : pt.x) - 1)
            res.y = pt.y + 1
            if(res.y == height) {
                res.y = 0
            }
        }

        return res
    }

    func canBuildRoad(at pt: MapPoint, boatRoad: Bool = false) -> Bool {
        if let object = object(at: pt), object.blockingType != .none {
            return false
        }

//        // dont build on the border
//        if(GetNode(pt).boundary_stones[0])
//            return false;

        for dir in Direction.allDirections {

            if let object = object(at: neighbor(of: pt, direction: dir)), object.blockingType == .nothingAround {
                return false;
            }

            // Other roads at this point?
            if hasRoad(at: pt, in: dir) {
                return false
            }
        }

//        // Terrain (unterscheiden, ob Wasser und Landweg)
//        if(!boat_road)
//        {
//            bool flagPossible = false;
//
//            for(unsigned char i = 0; i < 6; ++i)
//            {
//                TerrainBQ bq = GetDescription().get(GetRightTerrain(pt, Direction::fromInt(i))).GetBQ();
//                if(bq == TerrainBQ::DANGER)
//                    return false;
//                else if(bq != TerrainBQ::NOTHING)
//                    flagPossible = true;
//            }
//
//            return flagPossible;
//        } else
//        {
//            // Beim Wasserweg muss um den Punkt herum Wasser sein
//            if(!IsWaterPoint(pt))
//                return false;
//        }

        return true;
    }

    func hasRoad(at pt: MapPoint, in direction: Direction) -> Bool {
        if direction.isEastOrSouth {
            return road(at: pt, in: direction.reversed) != .none
        } else {
            return road(at: neighbor(of: pt, direction: direction), in: direction) != .none

        }
    }

    func road(at pt: MapPoint, in direction: Direction) -> Road.RoadType {
        return node(at: pt).roads[Int(direction.rawValue)]
    }

    func isPlayersTerritory(at pt: MapPoint) -> Bool {
        return true
    }
}

File: Unity/PathFindingFun/Assets/Scripts/World Objects/World.cs
using System;
using Unity;

public class World
{

    private readonly int width, height;
    private MapNode[] mapNodes;

    public World(int width, int height)
    {
        this.width = width;
        this.height = height;
        mapNodes = new MapNode[width * height];

    }

    private void initializeNodes()
    {
        for (int i = 0; i < width * height; i++)
        {
            mapNodes[i].roads = new Road.Types[] { Road.Types.none, Road.Types.none, Road.Types.none };
            mapNodes[i].buildQuality = BuildQuality.nothing;

        }
    }

    public int GetIndex(MapPoint pt)
    {
        return (pt.y * width) + pt.x;
    }

    public ref MapNode GetNodeRef(MapPoint pt)
    {
        return ref mapNodes[GetIndex(pt)];
    }

    public MapNode GetNode(MapPoint pt)
    {
        return mapNodes[GetIndex(pt)];
    }

    public MapObject GetObject(MapPoint pt)
    {
        return GetNode(pt).mapObject;
    }

    public void SetObject(MapPoint pt, MapObject mapObject)
    {
        GetNodeRef(pt).mapObject = mapObject;
    }

    public void SetPointRoad(MapPoint pt, Direction direction, Road.Types type)
    {
        if (Directions.isEastOrSouth(direction))
        {
            SetNodeRoad(pt, Directions.reversed(direction), type);
        }
        else
        {

        }
    }

    private void SetNodeRoad(MapPoint pt, Direction direction, Road.Types type)
    {
        GetNodeRef(pt).roads[(int)direction] = type;
    }

    private bool SetBuildQuality(MapPoint pt, BuildQuality bq)
    {
        BuildQuality oldBQ = GetNode(pt).buildQuality;
        GetNodeRef(pt).buildQuality = bq;
        return oldBQ != bq;
    }

    public int CalcDistance(MapPoint p1, MapPoint p2)
    {
        int dx = ((p1.x - p2.x) * 2) + (p1.y & 1) - (p2.y & 1);
        int dy = ((p1.y > p2.y) ? p1.y - p2.y : p2.y - p1.y) * 2;

        if (dx < 0)
            dx = -dx;

        if (dy > height)
        {
            dy = (height * 2) - dy;
        }

        if (dx > width)
        {
            dx = (width * 2) - dx;
        }

        dx -= dy / 2;

        return ((dy + (dx > 0 ? dx : 0)) / 2);
    }

    public MapPoint GetNeighbor(MapPoint pt, Direction direction)
    {
        /*  Note that every 2nd row is shifted by half a triangle to the left, therefore:
           Modifications for the dirs:
           current row:    Even    Odd
                        W  -1|0   -1|0
           D           NW  -1|-1   0|-1
           I           NE   0|-1   1|-1
           R            E   1|0    1|0
                       SE   0|1    1|1
                       SW  -1|1    0|1
        */

        int x;
        int y;
        switch (direction)
        {
            case Direction.west: // -1|0   -1|0
                x = ((pt.x == 0) ? width : pt.x) - 1;
                y = pt.y;
                break;
            case Direction.northwest: // -1|-1   0|-1
                x = (pt.y & 1) != 0 ? pt.x : (((pt.x == 0) ? width : pt.x) - 1);
                y = ((pt.y == 0) ? height : pt.y) - 1;
                break;
            case Direction.northeast: // 0|-1  -1|-1
                x = (!((pt.y & 1) != 0)) ? pt.x : ((pt.x == width - 1) ? 0 : pt.x + 1);
                y = ((pt.y == 0) ? height : pt.y) - 1;
                break;
            case Direction.east: // 1|0    1|0
                x = pt.x + 1;
                if (x == width)
                    x = 0;
                y = pt.y;
                break;
            case Direction.southeast: // 1|1    0|1
                x = (!((pt.y & 1) != 0)) ? pt.x : ((pt.x == width - 1) ? 0 : pt.x + 1);
                y = pt.y + 1;
                if (y == height)
                    y = 0;
                break;
            default: // 0|1   -1|1
                x = (pt.y & 1) != 0 ? pt.x : (((pt.x == 0) ? width : pt.x) - 1); //-V537
                y = pt.y + 1;
                if (y == height)
                    y = 0;
                break;
        }


        return new MapPoint((ushort)x, (ushort)y);
    }

    public MapPoint GetSecondNeighbor(MapPoint pt, int direction)
    {
        int ptx = (int)pt.x;
        int pty = (int)pt.y;


        int[] yShift = { 0, -1, -2, -2, -2, -1, 0, 1, 2, 2, 2, 1 };

        switch (direction)
        {
            case 0:
                ptx -= 2; break;
            case 1:
                ptx -= 2 - (((pt.y & 1) != 0) ? 1 : 0); break;
            case 2:
                ptx -= 1; break;
            case 3: break;
            case 4:
                ptx += 1; break;
            case 5:
                ptx += 2 - (((pt.y & 1) != 0) ? 0 : 1); break;
            case 6:
                ptx += 2; break;
            case 7:
                ptx += 2 - (((pt.y & 1) != 0) ? 0 : 1); break;
            case 8:
                ptx += 1; break;
            case 9: break;
            case 10: ptx -= 1; break;
            default:
                ptx -= 2 - (((pt.y & 1) != 0) ? 1 : 0); break;
        }

        pty += yShift[direction];

        if (direction > 3 && direction < 9)
        {
            ptx %= width;
        }

        if (direction > 6)
        {
            pty %= height;
        }

        if (ptx < 0)
        {
            ptx += width;
        }

        if (pty < 0)
        {
            pty += height;
        }
        return new MapPoint(ptx, pty);

    }

    public BuildQuality GetBuildQuality(MapPoint pt, int player)
    {
        BuildQuality nodeQuality = GetNode(pt).buildQuality;
        if (nodeQuality == BuildQuality.nothing || !isPlayerTerritory(pt, player))
        {
            return BuildQuality.nothing;
        }

        if (nodeQuality != BuildQuality.flag && !isPlayerTerritory(GetNeighbor(pt, Direction.southeast), player))
        {
            if (GetObject(GetNeighbor(pt, Direction.west))?.blockingType == BlockingType.flag)
            {
                return BuildQuality.nothing;
            }
            if (GetObject(GetNeighbor(pt, Direction.northwest))?.blockingType == BlockingType.flag)
            {
                return BuildQuality.nothing;
            }
            if (GetObject(GetNeighbor(pt, Direction.northeast))?.blockingType == BlockingType.flag)
            {
                return BuildQuality.nothing;
            }

            return BuildQuality.flag;
        }
        else
        {
            return nodeQuality;
        }

    }

    public bool CanBuildRoad(MapPoint pt, bool forBoat)
    {
        if (GetObject(pt)?.blockingType == BlockingType.none)
        {
            return false;
        }

        foreach (Direction dir in Directions.allDirections)
        {
            if (GetObject(GetNeighbor(pt, dir))?.blockingType == BlockingType.nothingAround)
            {
                return false;
            }
            if (HasRoad(pt, dir))
            {
                return false;
            }

        }

        return true;
    }

    public bool HasRoad(MapPoint pt, Direction dir)
    {
        if (Directions.isEastOrSouth(dir))
        {
            return GetRoad(pt, Directions.reversed(dir)) != Road.Types.none;
        }
        else
        {
            return GetRoad(GetNeighbor(pt, dir), dir) != Road.Types.none;
        }
    }

    public Road.Types GetRoad(MapPoint pt, Direction dir)
    {
        return GetNode(pt).roads[(int)dir];
    }

    public bool isPlayerTerritory(MapPoint pt, int player)
    {
        return true;
    }

    public bool IsOnRoad(MapPoint pt)
    {
        foreach(Direction roadDir in Directions.westNorth) {
            if (HasRoad(pt, roadDir) || HasRoad(GetNeighbor(pt, roadDir), roadDir)) {
                return true;
            }
        }


        return false;
    }

    public void RecalcBuildQualityForRoad(MapPoint pt)
    {
        RecalcBuildQuality(pt);
        RecalcBuildQuality(GetNeighbor(pt, Direction.east));
        RecalcBuildQuality(GetNeighbor(pt, Direction.southeast));
        RecalcBuildQuality(GetNeighbor(pt, Direction.southwest));

    }

    public void RecalcBuildQualityAround(MapPoint pt, bool extended = false)
    {
        RecalcBuildQuality(pt);

        foreach(Direction dir in Directions.allDirections)
        {
            RecalcBuildQuality(GetNeighbor(pt, dir));
        }

        if (extended) {
            for (int i = 0; i < 12; ++i)
            {
                RecalcBuildQuality(GetSecondNeighbor(pt, i));
            }
        }
        
    }


    public void RecalcBuildQuality(MapPoint pt)
    {
        BuildQualityCalculator calculator = new BuildQualityCalculator(this);
        if (SetBuildQuality(pt, calculator.GetBuildQuality(pt))) {
            UnityEngine.Debug.Log("Notify BQ Change!");
        }
    }

    public bool IsFlagAround(MapPoint pt) {
        foreach(Direction dir in Directions.allDirections) {
            if (GetObject(GetNeighbor(pt, dir))?.blockingType == BlockingType.flag) {
                return true;
            }
        }
        return false;
    }


    public void PlaceFlag(MapPoint pt, int player, Direction dir)
    {
        if (GetBuildQuality(pt, player) == BuildQuality.nothing) {
            UnityEngine.Debug.Log("Failed to place flag, BQ");
            return;
        }

        if (IsFlagAround(pt))
        {
            UnityEngine.Debug.Log("Failed to place flag, Flag Around");
            return;
        }

        if (GetObject(pt) is Flag)
        {
            UnityEngine.Debug.Log("Failed to place flag, already one there!");
            return;
        }

        SetObject(pt, new Flag(pt, player));
        RecalcBuildQualityAround(pt, true);
        
    }

}
